src/hooks/execution/SpendingLimitHook.sol

```diff
@@ -0,0 +1,240 @@
+import {DynamicArrayLib} from "solady/utils/DynamicArrayLib.sol";
+import {SafeTransferLib} from "solady/utils/SafeTransferLib.sol";
+import {IExecutionHook} from "../../interfaces/IExecutionHook.sol";
+import {AccountKeyHash, AccountKeyHashLib} from "../shared/AccountKeyHashLib.sol";
+
+enum SpendPeriod {
+    Minute,
+    Hour,
+    Day,
+    Week,
+    Month,
+    Year
+}
+
+interface ISpendingLimitHook is IExecutionHook {
+    /// @dev Exceeded the spend limit.
+    error ExceededSpendLimit();
+
+    /// @dev Emitted when a spend limit is set.
+    event SpendLimitSet(AccountKeyHash accountKeyHash, address token, SpendPeriod period, uint256 limit);
+
+    /// @dev Emitted when a spend limit is removed.
+    event SpendLimitRemoved(AccountKeyHash accountKeyHash, address token, SpendPeriod period);
+
+    function setSpendLimit(bytes32 keyHash, address token, SpendPeriod period, uint256 limit) external;
+}
+
+/// @title SpendingLimitHook
+/// @notice Example use case of an execution hook to set spending limits on specific key hashes.
+/// @author modified from https://github.com/ithacaxyz/account/blob/main/src/GuardedExecutor.sol
+contract SpendingLimitHook is ISpendingLimitHook {
+    using EnumerableSetLib for *;
+    using DynamicArrayLib for *;
+    using AccountKeyHashLib for bytes32;
+
+    ////////////////////////////////////////////////////////////////////////
+    // Structs
+    ////////////////////////////////////////////////////////////////////////
+
+    /// @dev Information about a spend.
+    /// All timestamp related values are Unix timestamps in seconds.
+    struct SpendInfo {
+        /// @dev Address of the token. `address(0)` denotes native token.
+        address token;
+        /// @dev The type of period.
+        SpendPeriod period;
+        /// @dev The maximum spend limit for the period.
+        uint256 limit;
+        /// @dev The amount spent in the last updated period.
+        uint256 spent;
+        /// @dev The start of the last updated period.
+        uint256 lastUpdated;
+        /// @dev The amount spent in the current period.
+        uint256 currentSpent;
+        /// @dev The start of the current period.
+        uint256 current;
+    }
+
+    /// @dev Holds the storage for the token period spend limits.
+    /// All timestamp related values are Unix timestamps in seconds.
+    struct TokenPeriodSpendStorage {
+        /// @dev The maximum spend limit for the period.
+        uint256 limit;
+        /// @dev The amount spent in the last updated period.
+        uint256 spent;
+        /// @dev The start of the last updated period (unix timestamp).
+        uint256 lastUpdated;
+    }
+
+    /// @dev Holds the storage for the token spend limits.
+    struct TokenSpendStorage {
+        /// @dev An enumerable set of the periods.
+        EnumerableSetLib.Uint8Set periods;
+        /// @dev Mapping of `uint8(period)` to `TokenPeriodSpendStorage`.
+        mapping(uint256 => TokenPeriodSpendStorage) spends;
```

Comment:
why is this BY period? How would that work if you have two spend periods for the same token that overlap? You'll have to check both periods states and combine the spends somehow?

Reply:
If you have two periods for the same token they all must be true:
- Minute, 10 AND hour, 100

You would loop through all of the periods, increment the spend for each and check none exceed?

Reply:
You add the same amount to each tracked period spend